src/Functions/FunctionTokens.h

```diff
@@ -117,32 +124,40 @@ class FunctionTokens : public IFunction
             ColumnString::Offset current_src_offset = 0;
             ColumnArray::Offset current_dst_offset = 0;
             ColumnString::Offset current_dst_strings_offset = 0;
-            for (size_t i = 0; i < size; ++i)
-            {
-                Pos pos = reinterpret_cast<Pos>(&src_chars[current_src_offset]);
-                current_src_offset = src_offsets[i];
-                Pos end = reinterpret_cast<Pos>(&src_chars[current_src_offset]) - 1;
-
-                generator.set(pos, end);
-                size_t j = 0;
-                while (generator.get(token_begin, token_end))
-                {
-                    size_t token_size = token_end - token_begin;
-
-                    res_strings_chars.resize(res_strings_chars.size() + token_size + 1);
-                    memcpySmallAllowReadWriteOverflow15(&res_strings_chars[current_dst_strings_offset], token_begin, token_size);
-                    res_strings_chars[current_dst_strings_offset + token_size] = 0;
-
-                    current_dst_strings_offset += token_size + 1;
-                    res_strings_offsets.push_back(current_dst_strings_offset);
-                    ++j;
-                }
-
-                current_dst_offset += j;
-                res_offsets.push_back(current_dst_offset);
+
+#define PROCESS_WITH_CUSTOM_RESIZE(RESIZE_METHOD) \
+            for (size_t i = 0; i < size; ++i) \
+            { \
+                Pos pos = reinterpret_cast<Pos>(&src_chars[current_src_offset]); \
+                current_src_offset = src_offsets[i]; \
+                Pos end = reinterpret_cast<Pos>(&src_chars[current_src_offset]) - 1; \
+                generator.set(pos, end); \
+                size_t j = 0; \
+                while (generator.get(token_begin, token_end)) \
+                { \
+                    size_t token_size = token_end - token_begin; \
+                    res_strings_chars.RESIZE_METHOD(res_strings_chars.size() + token_size + 1); \
```

Comment:
A macro isn't very beautiful here. Let's do it before the latest commits like

```cpp
if (res_chars_reserve_size < 0)
    res_strings_chars.resize(res_strings_chars.size() + token_size + 1);
else
    res_strings_chars.resize_assume_reserved(res_strings_chars.size() + token_size + 1);
```

Reply:
One path will do a reserve + resize, the other path will only do a resize. A reserve potentially does a heap allocation which is much more expensive than just a resize (that only adjusts `PodArrayBase::c_end`).

Yes, we are here in a hot loop but the costs are dominated by reserve operations. Additionally evaluating condition `res_chars_reserve_size < 0` will make no performance difference (but the code will be prettier).

Reply:
Pls refer to https://github.com/ClickHouse/ClickHouse/pull/62392#discussion_r1560384837